plutus-tx-plugin/test/Lift/Spec.hs

```diff
@@ -57,7 +54,6 @@ tests = testNested "Lift" [
     , goldenUPlc "list" (Lift.liftProgramDef ([1]::[Integer]))
     , goldenUEval "listInterop" [ getPlc listMatch, Lift.liftProgramDef ([1]::[Integer]) ]
     , goldenUPlc "nested" (Lift.liftProgramDef (NestedRecord (Just (1, 2))))
-    , goldenUPlc "bytestring" (Lift.liftProgramDef (WrappedBS "hello"))
```

Comment:
There is a problem with lifting `Builtins.ByteString` by `liftProgram` as `Lift` instances require `PLC.Includes`. And `PLC.Includes` instances are defined in `plutus-core` in `PlutusCore.Default.Universe`. And `plutus-tx` depends on `plutus-core`. That forbids lifting of builtin strings and bytestrings.

Should I try to provide the instances in `PlutusTx.ByteString.Instances` and `PlutusTx.String.Instances`? Or they don't have instances by design?

Reply:
Can't we define them here? https://github.com/input-output-hk/plutus/blob/master/plutus-tx/src/PlutusTx/Lift/Instances.hs

We did have an instance for bytestrings before. We didn't have one for strings, but I think that was just an oversight.

Reply:
I will try!

Reply:
It looks like it’s impossible at current state.

```
instance PLC.DefaultUni `PLC.Contains` ByteString where knownUni = DefaultUniByteString
```

doesn’t work because `DefaultUniByteString :: DefaultUni (Esc BS.ByteString)`.

Maybe we need some kind of an equality on type level? To make synonyms `Plutus.ByteString ~ BS.ByteString` and `Plutus.String ~ [Char]`?

Reply:
Um. This is a bit complicated. Let me write down my thoughts.

`typeRepBuiltin` and `liftBuiltin` work when we're lifting a type that is *actually* in the builtin universe. For things like `Integer`, that works perfectly, since we use the same `Integer` type in our source programs as we do for the actual builtin integer type in the universe.

But here (and with `String`) we deviate from that. We use actual Haskell `ByteString`s in the builtin universe, but we have a different opaque type in our programs. So the simple approach doesn't work.

The first thing I thought was: okay, let's write `instance Lift OpaqueByteString` that takes an opaque bytestring and turns it into a real bytestring inside a builtin bytestring constant. But our opaque type is actually empty! You can't make any useful values of that type in a _Haskell_ program.

So then I can see two ways forwards:

1. Make the opaque bytestring into a newtype around a real bytestring, maybe add `fromHaskellByteString :: ByteString -> OpaqueByteString1.

The downside of this is that `fromHaskellByteString` obviously won't work on chain, and it's a bit of a weird thing to do, you'll only do it just before you `Lift` something.

2. Write `instance Lift ByteString`! i.e. let people lift _real_ bytestrings, for which the `liftBuiltin` approach will actually work. People can then construct them however they want as normal.

The downside of this is that you slightly weirdly have two types that you have to work with. This would get really annoying if they're nested. How do you lift `data MyRecord = MyRecord { a :: OpaqueByteString }`? You can't! But if you make the field a real bytestring, it then won't work in your on-chain code. Argh!

---

Having written all that out, I think the only viable thing to do is approach 1. We'll have a slightly odd conversion function, but we can clearly say that it won't work on-chain.

Reply:
Okay, I will try the first approach.

But also note that there is already instance for `BS.ByteString`:

```
instance uni `PLC.Includes` BS.ByteString => Lift uni BS.ByteString where
    lift = liftBuiltin
```

Should we remove it then? By introducing `OpaqueByteString`.

And then there will real bytestrings and onchain bytestrigns wrapped with the newtype?

Reply:
Well, again, the problem is with `PLC.DefaultUni `PLC.Contains` ByteString`.

Reply:
Sorry, I didn't really mean a new type by `OpaqueByteString`, I just wanted a less-ambiguous way to refer to the `ByteString` defined in `Builtins`!

Then I think we can do something like this:
```
type HByteString = BS.ByteString
type PByteString = Builtins.ByteString -- the actual definition is a newtype around BS.ByteString

fromHaskellByteString :: HByteString -> PByteString -- just applies the newtype constructor

DefaultUni `Contains` HByteString -- this remains t rue

-- instances for PByteString are implemented in terms of the underlying HByteString
instance uni `PLC.Includes` HByteString => Typeable uni PByteString where
    typeRep = typeRepBuiltin @HByteString

instance uni `PLC.Includes` HByteString => Lift uni PByteString where
    lift (PByteString bs) = liftBuiltin bs
```

Reply:
Aha, `Lift` instance works this way, but `Typeable` doesn't.

```
src/PlutusTx/Lift/Instances.hs:96:15: error:
    • Couldn't match type ‘ByteString’
                     with ‘BS.ByteString’
      NB: ‘BS.ByteString’
            is defined in ‘Data.ByteString.Internal’
                in package ‘bytestring-0.10.12.0’
          ‘ByteString’
            is defined in ‘PlutusTx.ByteString.Internal’
                in package ‘plutus-tx-0.1.0.0’
      Expected type: Proxy ByteString
                     -> RTCompile uni fun (Type TyName uni ())
        Actual type: Proxy BS.ByteString
                     -> RTCompile uni fun (Type TyName uni ())
    • In the expression: typeRepBuiltin @BS.ByteString
      In an equation for ‘typeRep’:
          typeRep = typeRepBuiltin @BS.ByteString
      In the instance declaration for ‘Typeable uni ByteString’
   |
96 |     typeRep = typeRepBuiltin @BS.ByteString
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Is only `Lift` enough? (It's certainly enough for plugin tests)

And should we do the same trick for `String`? To make it a newtype around `[Char]`?

Reply:
Ah sorry, I got it wrong, I thought that `typeRep` took just a type application but in fact it takes a proxy, so you'll need something like `typeRep _proxyPByteString = typeRepBuiltin (Proxy @HByteString)`

We'll need `Typeable` for types that contain bytestrings, because then we start having to mention the types of the field in the generated program.

> And should we do the same trick for String? To make it a newtype around [Char]?

Probably - but let's see if this approach works at all first!

Reply:
> `typeRep _proxyPByteString = typeRepBuiltin (Proxy @HByteString)`

It compiles! I will try to adopt the code to the newtype then.

Reply:
Yeah hopefully this works. I'm still not totally sure, we'll have to see :sweat_smile: